### PR TITLE
Check all documents specified in Product_Document tags for correct mi…

### PR DESCRIPTION
**Pull request by qchaupds**
…me types.

1. Add new test resources :src/test/resources/github367
2. Add needed dependency for group com.sun.activation, which is the implementation jar :pom.xml
3. Add default mime type resource file :src/main/resources/validate_default_mime_types.txt
4. Add new class to validate JPEG and PNG image file content :src/main/java/gov/nasa/pds/tools/util/ImageUtil.java
5. Add new class to check if the mime type of a given file matches its file extension :src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java
6. Add new problem types :src/main/java/gov/nasa/pds/tools/validate/ProblemType.java
7. Add check for all file_name tags in Product_Document tags :src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java
8. Add new regression test :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/367 As a user, I want to validate all files referenced by a Product_Document

Previously, validate module only check for the correct file format of PDF.  Now, the code checks for all file names
for matching mime types.  Code also check for correctness of JPEG and PNG file content.
Note that this does not check for correct file content for the other un-mentioned types.

Tested in DEV:

Case 1:
Running validate against a bundle with document label with mime types incorrect.

```
validate -R pds4.bundle -r report_github367_bundle_valid.json  -s json -t src/test/resources/github367/bundle_kaguya_derived.xml >& t3
```

There should be 21 warning messages, with 12 messages related to the wrong mime types.

```
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 213 % grep "not correct" report_github367_bundle_valid.json | cat -n
     1            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_2.txt with document standard id 'Microsoft Word' is not correct"
     2            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.html with document standard id 'MPEG-4' is not correct"
     3            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.ps with document standard id 'HTML' is not correct"
     4            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.eps with document standard id 'TIFF' is not correct"
     5            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.gif with document standard id 'TIFF' is not correct"
     6            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.doc with document standard id 'JPEG' is not correct"
     7            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.tex with document standard id 'GIF' is not correct"
     8            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.mp4 with document standard id 'PDF' is not correct"
     9            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.xls with document standard id 'Encapsulated Postscript' is not correct"
    10            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.doc with document standard id 'HTML' is not correct"
    11            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.rtf with document standard id 'LaTEX' is not correct"
    12            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.tif with document standard id 'Microsoft Excel' is not correct"
```

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/util/ImageUtil.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+   * @param pngBase The basename of the PNG file
+   * @return true if the PNG file is valid false otherwise
+   *
+   */
+  public boolean isPNG(String pngBase) throws Exception {
+
+    LOG.debug("isPNG:pngBase {}",pngBase);
+
+    // Get the location of the input file.
+    URI uri = null;
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ImageUtil.java:142]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/ImageUtil.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    String pngRef = parent + File.separator + pngBase;
+    LOG.debug("isPNG:parent,pngBase,pngRef [{}],[{}],[{}]",parent,pngBase,pngRef);
+
+    File pngFile = new File(pngRef);
+    RandomAccessFile raf = new RandomAccessFile(pngFile, "r");
+
+    try {
+        // Read the first 8 bytes.
+        int firstInt = 0;
+        int secondInt = 0;
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ImageUtil.java]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/ImageUtil.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    String jpegRef = parent + File.separator + jpegBase;
+    LOG.debug("isJPEG:parent,jpegBase,jpegRef [{}],[{}],[{}]",parent,jpegBase,jpegRef);
+
+    File jpegFile = new File(jpegRef);
+    RandomAccessFile raf = new RandomAccessFile(jpegFile, "r");
+
+    try {
+        // Read the first 2 bytes and last 2 bytes.  If they are 0xffd8 and 0xffd9 then the file is a JPEG file.
+        int firstShort = 0; 
+        int lastShort = 0;
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ImageUtil.java]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/ImageUtil.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    // Build the full pathname of the file.
+    String jpegRef = parent + File.separator + jpegBase;
+    LOG.debug("isJPEG:parent,jpegBase,jpegRef [{}],[{}],[{}]",parent,jpegBase,jpegRef);
+
+    File jpegFile = new File(jpegRef);
+    RandomAccessFile raf = new RandomAccessFile(jpegFile, "r");
+
+    try {
+        // Read the first 2 bytes and last 2 bytes.  If they are 0xffd8 and 0xffd9 then the file is a JPEG file.
+        int firstShort = 0;
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ImageUtil.java]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/ImageUtil.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    // Build the full pathname of the file.
+    String pngRef = parent + File.separator + pngBase;
+    LOG.debug("isPNG:parent,pngBase,pngRef [{}],[{}],[{}]",parent,pngBase,pngRef);
+
+    File pngFile = new File(pngRef);
+    RandomAccessFile raf = new RandomAccessFile(pngFile, "r");
+
+    try {
+        // Read the first 8 bytes.
+        int firstInt = 0;
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/ImageUtil.java]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java [hunk 6]
```diff
@@ -350,6 +426,52 @@ private boolean validate(DocumentInfo xml) {
...
+                      ExceptionType.ERROR, ProblemType.INTERNAL_ERROR,
+                      "Error occurred while processing PNG file content for "
+                      + FilenameUtils.getName(urlRef.toString()) + ": "
+                      + e.getMessage());
+                  problems.add(new ValidationProblem(def, target,
+                      fileObject.getLineNumber(), -1));
+                  passFlag = false;
+                }
+
+                passFlag = this.checkGenericDocument(target, urlRef, fileObject, htmlName, parent, directory, documentStandardId, "HTML", problems, ProblemType.NON_HTML_FILE);
```
Comment (jordanpadams) [inline src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java:452]: @qchaupds it seems like this should be a loop through a hashmap, but not worth holding up the PR right now.

**Review by jordanpadams — APPROVED** [review]

**Review by jordanpadams — CHANGES_REQUESTED** [review]
@qchaupds looks like these tests are failing. can you please fix ASAP and I will merge?

File: src/main/java/gov/nasa/pds/tools/util/MimeTable.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+              tokenIndex = 1;
+              while (tokenIndex < tokens.length) {
+                  LOG.debug("loadMimeTable:APPENDING_KEY {},{},[{}],[{}]",tokenIndex,tokens.length,tokens[tokenIndex],tokens[0]);
+                  mimetypesFileTypeMap.put(tokens[tokenIndex],tokens[0]);  // Each file extension becomes a key and the first token (mime type) becomes the value
+                  tokenIndex += 1;
+              }
+             
+          }
+      }
+      LOG.debug("loadMimeTable:defaultMimeTypesFileName,mimetypesFileTypeMap.size {},{}",defaultMimeTypesFileName,mimetypesFileTypeMap.size());
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/MimeTable.java:83]: @qchaupds let's definitely check this leak if we can.

File: src/main/java/gov/nasa/pds/tools/util/MimeTable.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+      // which is missing the leading slash.
+      //
+
+      Scanner inFile = null;
+      try {
+          String parent = defaultMimeTypesFileName.substring(0,defaultMimeTypesFileName.lastIndexOf(File.separator));
+          LOG.debug("loadMimeTable:FilenameUtils.getName(defaultMimeTypesFileName) {}",FilenameUtils.getName(defaultMimeTypesFileName));
+          LOG.debug("loadMimeTable:FilenameUtils.getPath(defaultMimeTypesFileName) {}",FilenameUtils.getPath(defaultMimeTypesFileName));
+          LOG.debug("loadMimeTable:parent {}",parent);
+          inFile = new Scanner(new File(parent + File.separator + FilenameUtils.getName(defaultMimeTypesFileName)));
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/util/MimeTable.java:53]: @sonatype-lift ignore  The name has already been sanitized by fetching the parent plus the filename separately.

**Review by jordanpadams — APPROVED** [review]

File: src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  private static final String MSWORD_ENDS_WITH_PATTERN_1 = ".doc";
+  private static final String MSWORD_ENDS_WITH_PATTERN_2 = ".docx";
+  private static final String MSEXCEL_ENDS_WITH_PATTERN_1 = ".xls";
+  private static final String MSEXCEL_ENDS_WITH_PATTERN_2 = ".xlsx";
+  private static final String LATEX_ENDS_WITH_PATTERN_1 = ".tex";
+  private static final String LATEX_ENDS_WITH_PATTERN_2 = ".TEX";
+  private static final String POSTSCRIPT_ENDS_WITH_PATTERN_1 = ".ps";
+  private static final String POSTSCRIPT_ENDS_WITH_PATTERN_2 = ".PS";
+  private static final String ENCAPSULATED_POSTSCRIPT_ENDS_WITH_PATTERN_1 = ".eps";
+  private static final String ENCAPSULATED_POSTSCRIPT_ENDS_WITH_PATTERN_2 = ".EPS";
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java:46]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/MimeTable.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+   * Given a default mime types table, load the table into a HashMap.
+   * @param defaultMimeTypesFileName The file name of the default mime types table 
+   * @return None
+   */
+
+  public void loadMimeTable(String defaultMimeTypesFileName) throws IOException {
+      LOG.debug("loadMimeTable:defaultMimeTypesFileName {}",defaultMimeTypesFileName);
+      Scanner inFile = null;
+      try {
+          inFile = new Scanner(new File(defaultMimeTypesFileName));
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/MimeTable.java]: @sonatype-lift help

File: src/main/java/gov/nasa/pds/tools/util/MimeTable.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+      while (inFile.hasNext()) {
+          nextLine = inFile.nextLine();
+          if (!nextLine.startsWith("#")) {
+              LOG.debug("loadMimeTable:APPENDING_LINE [{}]",nextLine);
+              // Parse the line for the mime type and the file extension.
+              // Each file extension becomes a key and the first token (mime type) becomes the value
+              // 7-Bit_ASCII_Text txt text TXT TEXT
+              tokens = nextLine.split("\\s+");
+
+              LOG.debug("loadMimeTable:nextLine,tokens.length [{}],{}",nextLine,tokens.length);
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/MimeTable.java]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/util/MimeTable.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+              tokenIndex = 1;
+              while (tokenIndex < tokens.length) {
+                  LOG.debug("loadMimeTable:APPENDING_KEY {},{},[{}],[{}]",tokenIndex,tokens.length,tokens[tokenIndex],tokens[0]);
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/util/MimeTable.java:75]: @sonatype-lift ignore

**Pull request merged**